src/mlpack/methods/decision_tree/best_binary_categorical_split_impl.hpp

```diff
@@ -0,0 +1,414 @@
+  uvec categoryRank(numCategories);
+  for (size_t i = 0; i < numCategories; i++)
+    categoryRank[sortedCategories[i]] = i;
+
+  uvec transformedData(n); 
+  for (size_t i = 0; i < n; ++i)
+    transformedData[i] = categoryRank[data[i]]; 
+
+  // Split the transformed vₖ as a numeric type.
+  bestFoundGain = NumericSplit::template SplitIfBetter<UseWeights>(
+      bestFoundGain,
+      transformedData, 
+      responses,
+      weights,
+      minLeafSize,
+      minGainSplit,
+      splitInfo,
+      (NumericAux &) aux,
+      fitnessFunction
+  );
+  bool improved = bestFoundGain != DBL_MAX;
+
+  if (improved)
+  {
+      // This split is better: store the set membership in splitInfo
+      // and return. Thus splitInfo is a vector of size Q, where Q is 
+      // the number of categories, and splitInfo[k] is zero if category 
+      // k is assigned to the left child, and otherwise it is one if k
+      // is assigned to the right.
+      size_t splitIndex = (size_t) std::floor((double) splitInfo[0]);
+      splitInfo.set_size(numCategories + 1);
+      for (size_t c: sortedCategories.subvec(0, splitIndex))
+        splitInfo[c] = LEFT; 
+      for (size_t c: sortedCategories.subvec(splitIndex+1, numCategories-1))
+        splitInfo[c] = RIGHT; 
+      return bestFoundGain;
+  }
+  return DBL_MAX;
+}
+
+template<typename FitnessFunction>
+template<typename VecType, typename LabelsType>
+bool BestBinaryCategoricalSplit<FitnessFunction>::PartitionSplit(
+    const VecType& data,
+    const LabelsType& labels,
+    const size_t numCategories,
+    const size_t numClasses,
+    double& bestFoundGain,
+    SpMat<short>& categorySamples,
+    uvec& categories,
+    vec& splitInfo,
+    Mat<size_t>& classCounts,
+    size_t totalLeft,
+    size_t totalRight,
+    size_t k)
+{
+  /* Base case -- We have already found an optimal split. */
+  if (bestFoundGain >= 0.0)
+    return false;
+
+  if (k == numCategories)
+  {
+    /**
+     * Base case -- Compute the gain for the current partition.
+     */
+    double leftGain, rightGain, gain;
+
+    // The gain for children Tₗ and Tᵣ. 
+    leftGain = FitnessFunction::template EvaluatePtr<false>(
+        classCounts.colptr(LEFT), numClasses, totalLeft
+    );
+    rightGain = FitnessFunction::template EvaluatePtr<false>(
+        classCounts.colptr(RIGHT), numClasses, totalRight
+    );
+    // The gain for this split. 
+    gain = double(totalLeft)*leftGain + double(totalRight)*rightGain;
+
+    // This is the best split found thus far. 
+    if (gain > bestFoundGain)
+    {
+      bestFoundGain = gain;
+      splitInfo.set_size(numCategories);
+      for (size_t i = 0; i < numCategories; ++i)
+        splitInfo[i] = categories[i];
+      return true;
+    }
+    return false;
+  }
+  bool improved = false;
+  uvec samples = find(categorySamples.col(k));
+
+  /**
+   * Compute the gain with category Cₖ ∈ Tₗ
+   */
+  categories[k] = LEFT;
+  for (auto i : samples)
```

Comment:
I know it makes work for you, but please don't use `auto`; it can have strange interactions with Armadillo (yes, in certain cases, even here, it has to do with the return types Armadillo provides), and the code is more readable for maintainers coming along later.

Reply:
Ok, no problem. I have written quite a bit of C, but I'm relatively new to C++. The `auto` type seemed strange to me, but I noticed that Stroustrup uses it all over the place in _Tour of C++_.